fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java

```diff
@@ -0,0 +1,116 @@
+import com.ibm.fhir.model.resource.SearchParameter;
+import com.ibm.fhir.model.type.code.ResourceTypeCode;
+import com.ibm.fhir.model.type.code.SearchParamType.Value;
+import com.ibm.fhir.search.util.SearchHelper;
+
+public class IncludesUtil {
+    private static final Logger LOG = java.util.logging.Logger.getLogger(IncludesUtil.class.getName());
+
+    /**
+     * Computes the list of supported include parameter values that target resource types from the passed resourcesToInclude list.
+     *
+     * @param baseResourceType
+     * @param resourceTypesToInclude
+     * @param searchHelper
+     * @return The list of values to use for the _include search parameter.
+     * @throws Exception
+     */
+    public static List<String> computeIncludesParamValues(String baseResourceType, List<String> resourceTypesToInclude, SearchHelper searchHelper)
+            throws Exception {
+        List<String> result = new ArrayList<>();
+
+        List<String> allowedIncludes = FHIRConfigHelper.getSearchPropertyRestrictions(baseResourceType, FHIRConfigHelper.SEARCH_PROPERTY_TYPE_INCLUDE);
+        Map<ResourceType, List<String>> codesByType = IncludesUtil.getSearchCodesByType(baseResourceType, searchHelper);
+
+        for (Map.Entry<ResourceType, List<String>> entry : codesByType.entrySet()) {
+            if (resourceTypesToInclude.contains(entry.getKey().value())) {
+                result.addAll(computeIncludes(baseResourceType, entry.getKey(), allowedIncludes, entry.getValue()));
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * @param compartmentMemberType
+     * @param searchHelper
+     * @return a map from resource target type to the list of reference search parameter codes that can target that type
+     * @throws Exception
+     */
+    static Map<ResourceType, List<String>> getSearchCodesByType(String compartmentMemberType, SearchHelper searchHelper) throws Exception {
+        Map<ResourceType, List<String>> codesByType = new HashMap<>();
+
+        Map<String, SearchParameter> allSearchParameters = searchHelper.getSearchParameters(compartmentMemberType);
+        for (Map.Entry<String, SearchParameter> entry : allSearchParameters.entrySet()) {
+            SearchParameter sp = entry.getValue();
+            if (sp.getType().getValueAsEnum() != Value.REFERENCE) {
+                continue;
+            }
+
+            String expression = sp.getExpression().getValue();
+            for (ResourceTypeCode target : sp.getTarget()) {
+                // The search parameter target types come from the possible types of the target reference
+                // and don't take into account any "is X" filters from within the expression,
+                // so we use this heuristic to avoid adding the code when expression filtering applies
+                // and does not include a filter clause for this particular target type.
+                if (expression.contains(".is(") || expression.contains(" is ")) {
```

Comment:
Is this brittle?

Reply:
its definitely "heuristic".  IMHO the conformance artifacts should account for this when they list the allowed "target" types for a search parameter of type reference.

Reply:
a potential alternative would be for us to patch the official resources in order to make those right...then we wouldn't need to do this.  but it wouldn't necessarily help with user-packaged IGs that may have the same issue

Reply:
I don't think it's a blocker...just struck me that the test could hit false positives...but I don't know enough about the context.

Reply:
ok.  i think its pretty safe.  theoretically someone could construct a FHIRPath to confuse us (like by combining an “all” or “iff” or “not” or something like that with these “is” clauses), but in practice no one does that for search parameters.

Its almost always just a simple path and sometimes they have an is at the end to filter to a particular type.